kernel/kernel-api/src/main/java/io/delta/kernel/internal/icebergcompat/IcebergWriterCompatV1MetadataValidatorAndUpdater.java

```diff
@@ -0,0 +1,217 @@
+              IN_COMMIT_TIMESTAMP_W_FEATURE,
+              // TODO: add clustering once we support it in Kernel
+              TIMESTAMP_NTZ_RW_FEATURE,
+              TYPE_WIDENING_RW_FEATURE,
+              TYPE_WIDENING_PREVIEW_TABLE_FEATURE)
+          .collect(toSet());
+
+  /** Checks that all features supported in the protocol are in {@link #ALLOWED_TABLE_FEATURES} */
+  private static final IcebergCompatCheck UNSUPPORTED_FEATURES_CHECK =
+      (inputContext) -> {
+        // TODO include additional information in inputContext so that we can throw different errors
+        //   for blocking enablement of icebergWriterCompatV1 versus enablement of the incompatible
+        //   feature
+        if (!ALLOWED_TABLE_FEATURES.containsAll(
+            inputContext.newProtocol.getImplicitlyAndExplicitlySupportedFeatures())) {
+          Set<TableFeature> incompatibleFeatures =
+              inputContext.newProtocol.getImplicitlyAndExplicitlySupportedFeatures();
+          incompatibleFeatures.removeAll(ALLOWED_TABLE_FEATURES);
+          throw DeltaErrors.icebergWriterCompatV1IncompatibleTableFeatures(incompatibleFeatures);
+        }
+      };
+
+  /**
+   * Checks that there are no unsupported types in the schema. Data types {@link ByteType} and
+   * {@link ShortType} are unsupported for IcebergWriterCompatV1 tables.
+   */
+  private static final IcebergCompatCheck UNSUPPORTED_TYPES_CHECK =
+      (inputContext) -> {
+        List<Tuple2<List<String>, StructField>> matches =
+            SchemaUtils.filterRecursively(
+                inputContext.newMetadata.getSchema(),
+                /* recurseIntoMapAndArrayTypes= */ true,
+                /* stopOnFirstMatch = */ false,
+                field -> {
+                  DataType dataType = field.getDataType();
+                  return (dataType instanceof ByteType || dataType instanceof ShortType);
+                });
+
+        if (!matches.isEmpty()) {
+          throw DeltaErrors.icebergCompatUnsupportedTypeColumns(
+              INSTANCE.compatVersion(),
+              matches.stream().map(tuple -> tuple._2.getDataType()).collect(toList()));
+        }
+      };
+
+  /**
+   * Checks that the table feature `invariants` is not active in the table, meaning there are no
+   * invariants stored in the table schema.
+   */
+  private static final IcebergCompatCheck INVARIANTS_INACTIVE_CHECK =
+      // Note - since Kernel currently does not support the table feature `invariants` we will not
+      // hit this check for E2E writes since we will fail early due to unsupported write
+      // If Kernel starts supporting the feature `invariants` this check will become applicable
+      (inputContext) -> {
+        if (TableFeatures.hasInvariants(inputContext.newMetadata.getSchema())) {
+          throw DeltaErrors.icebergWriterCompatV1IncompatibleTableFeatures(
+              Collections.singleton(INVARIANTS_W_FEATURE));
+        }
+      };
+
+  @Override
+  String compatVersion() {
```

Comment:
nit: the "version" is actually just the number. I think this name was already around, but if it's not too much work, could we call this `compatFeatureName` or similar. I think it'll get confusing when we have more versions to deal with.

Reply:
yep that's easy to fix I will update it